lib/elixir/lib/calendar/date_range.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Date.Range do
+  @moduledoc """
+  Defines a range of dates.
+
+  A range of dates represents a discrete number of dates where
+  the first and last values are dates with matching calendars.
+
+  Ranges of dates can be either increasing (`first <= last`) or
+  decreasing (`first > last`). They are also always inclusive.
+
+  A range of dates implements the `Enumerable` protocol, which means
+  functions in the `Enum` module can be used to work with
+  ranges:
+
+      iex> range = Date.range(~D[2001-01-01], ~D[2002-01-01])
+      iex> Enum.count(range)
+      366
+      iex> Enum.member?(range, ~D[2001-02-01])
+      true
+      iex> Enum.reduce(range, 0, fn(_date, acc) -> acc - 1 end)
+      -366
+
+  """
+
+  @opaque t :: %__MODULE__{first: Date.t, last: Date.t}
```

Comment:
I'm not sure we want anyone accessing or modifying this fields, so I wouldn't add `first/1` and `last/1`.

Even if the `inspect` output looks like `#DateRange<first, last>`, it's just a `struct` so you could still use `Map./put/3` if you really needed to modify stuff.
And, since it's just a `struct` under the hood, you can still read the values with `range.first`.